File: torch/legacy/nn/SoftPlus.py
import torch
from .Module import Module


class SoftPlus(Module):

    def __init__(self, beta=1, threshold=20):
        super(SoftPlus, self).__init__()
        self.beta = beta              # Beta controls sharpness of transfer function
        self.threshold = threshold    # Avoid floating point issues with exp(x), x>20

    def updateOutput(self, input):
        # f(x) = 1/beta * log(1 + exp(beta * x))
        self._backend.SoftPlus_updateOutput(
            self._backend.library_state,
            input,
            self.output,
            self.beta,
            self.threshold
        )
        return self.output

    def updateGradInput(self, input, gradOutput):
        # d/dx[log(1+exp(k*x))/k] = exp(kx) / (exp(kx) + 1)
        # SINCE
        # y = (1/k)*log(1+exp(k*x)) #> x = (1/k)*log(exp(k*y)-1)
        # THEREFORE:
        # d/dx(f(x)) = (exp(k*y) - 1) / exp(k*y)
        self._backend.SoftPlus_updateGradInput(
            self._backend.library_state,
            input,
            gradOutput,
            self.gradInput,
            self.output,
            self.beta,
            self.threshold
        )
        return self.gradInput

File: aten/src/THNN/generic/SoftPlus.c
#ifndef TH_GENERIC_FILE
#define TH_GENERIC_FILE "THNN/generic/SoftPlus.c"
#else

void THNN_(SoftPlus_updateOutput)(
          THNNState *state,
          THTensor *input,
          THTensor *output,
          accreal beta_,
          accreal threshold_)
{
  scalar_t beta = TH_CONVERT_ACCREAL_TO_REAL(beta_);
  scalar_t threshold = TH_CONVERT_ACCREAL_TO_REAL(threshold_);
  THTensor_(resizeAs)(output, input);

  // f(x) = 1/beta * log(1 + exp(beta * x))
  TH_TENSOR_APPLY2(scalar_t, output, scalar_t, input,               \
    *output_data = (*input_data * beta) > threshold ? *input_data : std::log1p(exp(*input_data * beta)) / beta;
  );
}

void THNN_(SoftPlus_updateGradInput)(
          THNNState *state,
          THTensor *input,
          THTensor *gradOutput,
          THTensor *gradInput,
          THTensor *output,
          accreal beta_,
          accreal threshold_)
{
  scalar_t beta = TH_CONVERT_ACCREAL_TO_REAL(beta_);
  scalar_t threshold = TH_CONVERT_ACCREAL_TO_REAL(threshold_);
  THNN_CHECK_NELEMENT(input, gradOutput);
  THTensor_(resizeAs)(gradInput, output);

  // d/dx[log(1+exp(k*x))/k] = exp(kx) / (exp(kx) + 1)
  // SINCE
  // y = (1/k)*log(1+exp(k*x)) --> x = (1/k)*log(exp(k*y)-1)
  // THEREFORE:
  // d/dx(f(x)) = (exp(k*y) - 1) / exp(k*y)
  TH_TENSOR_APPLY3(scalar_t, gradInput, scalar_t, gradOutput, scalar_t, output,
    scalar_t z = exp(*output_data * beta);
    *gradInput_data = (*output_data * beta) > threshold ? *gradOutput_data : *gradOutput_data * (z - 1.)/z;
  );
}

#endif
